asmcomp/cmmgen.ml

```diff
@@ -300,12 +300,50 @@ let invert_then_else = function
   | Then_false_else_true -> Then_true_else_false
   | Unknown -> Unknown
 
+let rec simplify_cond = function
+  | Cop(Ccmpi cmpop
+       , [ Cop(Caddi,[Cop(Clsl,[left;Cconst_int(1,_)],_);Cconst_int(1,_)],_)
+         ; Cconst_int (right_i,right_dbg)
+         ]
+       , dbg_cmp
+       ) ->
+      Cop (Ccmpi cmpop, [ left; Cconst_int (right_i asr 1, right_dbg)], dbg_cmp)
+      |> simplify_cond
+  | Cop(Ccmpi cmpop
+       , [ Cconst_int (left_i,left_dbg)
+         ; Cop(Caddi, [Cop(Clsl,[right;Cconst_int(1,_)],_);Cconst_int(1,_)],_)
+         ]
+       , dbg_cmp
+       ) ->
+      Cop (Ccmpi cmpop, [ Cconst_int (left_i asr 1, left_dbg); right], dbg_cmp)
+      |> simplify_cond
+  | Cop (Ccmpi (Ceq | Cne as eqop)
+        , ( ([ Cop(Ccmpi cmpop, [ left; right ], dbg_cmp)
+             ; Cconst_int ((0 | 1 as true_false), _)
+             ])
+          | ([ Cconst_int ((0 | 1 as true_false), _)
+             ; Cop(Ccmpi cmpop, [ left; right ], dbg_cmp)
+             ])
+          )
+        , _dbg
+        ) ->
+      (* Simplify comparing the result of a comparison with true/false *)
+      let cmpop =
+        match eqop, true_false with
+        | Cne,0 | Ceq,1 -> cmpop (* <> false, = true *)
+        | Ceq,0 | Cne,1 -> negate_integer_comparison cmpop (* = false, <> true*)
+        | _ -> assert false
+      in
+      Cop (Ccmpi cmpop, [ left; right ], dbg_cmp)
+  | cond -> cond
+
 let mk_if_then_else dbg cond ifso_dbg ifso ifnot_dbg ifnot =
   match cond with
```

Comment:
We could use `match simplify_cond cond with` here. I *think* that it is equivalent to what you are doing today as simplify_cond will never turn a non-constant input into a constant output, but this may change.